src/nxt_runtime.c

```diff
@@ -797,6 +802,60 @@ nxt_runtime_conf_init(nxt_task_t *task, nxt_runtime_t *rt)
     return NXT_OK;
 }
 
+static nxt_int_t
+nxt_runtime_creds(nxt_task_t *task,  nxt_runtime_t *rt)
+{
+    char          *user, *group_name;
+    struct group  *grp;
+    struct passwd *pwd;
+
+    if (rt->capabilities.setid) {
+        if (rt->user_cred.user == NULL) {
+            rt->user_cred.user = NXT_USER;
+        }
+
+        if (rt->group == NULL) {
+            rt->group = NXT_GROUP;
+        }
+    } else {
+        pwd = getpwuid(geteuid());
+
+        if (pwd == NULL) {
+            user = group_name = NULL;
+        } else {
+            user = pwd->pw_name;
+            grp = getgrgid(pwd->pw_gid);
+
+            if (grp != NULL) {
+                group_name = grp->gr_name;
+            } else {
+                group_name = NULL;
+            }
+        }
+
+        if (user == NULL) {
+            nxt_alert(task, "Unit is unable to get the current username. "
+                "There's no entry for uid %d in passwd",
+                geteuid());
+            return NXT_ERROR;
+        }
+
+        if (rt->user_cred.user != user || rt->group != group_name) {
+            nxt_log(task, NXT_LOG_NOTICE, "Unit is running unprivileged, then it"
+                " cannot use arbitrary user and group. Using \"%s\" and \"%s\""
+                " for user and group, respectively.", user,
+                group_name ? group_name : "");
+
+            nxt_capability_log_hint(task);
+        }
+
+        rt->user_cred.user = user;
+        rt->group = group_name;
```

Comment:
Why don't just set `rt->group` to `NULL` here? The `nxt_user_cred_get()` function sets `uc->base_gid = pwd->pw_gid;` in this case.

Instead you're trying to get group name out of `pwd->pw_gid` and then pass this name to `nxt_user_cred_get()` where it converts back to GID (i.e. `pwd->pw_gid`).

I don't think that `nxt_user_cred_get()` need to be called at all. You just need to set `rt->user_cred.uid` to `pwd->pw_uid` and `rt->user_cred.base_gid` to `pwd->pw_gid`. There is no reason to convert gid/uid to names and then reverse.

Reply:
yes, that's odd but I called `nxt_user_cred_get` because it does additional stuff (that I'm not sure why, like calling `nxt_user_groups_get` and `getgroups`. I don't understand all use cases involved here, sorry.

Reply:
I think you can call `nxt_user_groups_get()` too by ether moving your function to `nxt_process.c` or making `nxt_user_groups_get()` public.

I'll ask @igorsysoev tomorrow why `nxt_user_groups_get()` is needed. Maybe it's not needed at all.

Reply:
@VBart I tried a different approach now, much simpler. Let me know if makes sense. 

If process has capabilities, then it just follow the old behaviour. If it does not, then we just warn the user that Unitd cannot setuid/setgid and we just ignore the creds.

If user configures an app passing a user different than the value in `rt->user_cred`, then it fails.

Reply:
ok, looks good.